### PR TITLE
New Steel Crest Ability: Soak

### PR DESCRIPTION
**Pull request by DecemberIsHere**
# About the pull request

This adds a new ability for the defender strain, Steel Crest called Soak. It is an ability that when activated, tracks damaged taken and once it reaches a threshold slightly heal the Steel Crest defender and reset the cooldown of its Tail Slam ability. The damage threshold needed to reach this state is 140 damage within 6 seconds, and will not do anything if it doesn't reach this cooldown.

# Explain why it's good for the game

Steel crest as a strain is extremely underwhelming, it's almost entirely dedicated to simply soaking damage. Its combat capabilities is withheld to headbutting from a corner and then going back in hiding. Its tail slam is underutilized as a result, and this PR will switch up gameplay and allow for steel crest to be a much more fun mix between its tanking ability, and its damage dealing aggressiveness. 

# Testing Photographs and Procedure

<details>
Tested against being hurt by a sentry while its active, it works. Will refresh your tail slam and slightly heal.
</details>

# Changelog
:cl: Usnpeepoo
add: Added a new ability to the Steel Crest strain, called Soak. It tracks your accumulated damage and once it reaches a threshold, reset your tail slam cooldown and slightly heal you.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_abilities.dm [hunk 1]
```diff
@@ -48,3 +48,16 @@
...
+	action_icon_state = "soak"
+	ability_name = "soak"
+	macro_path = /datum/action/xeno_action/verb/verb_soak
+	action_type = XENO_ACTION_ACTIVATE
+	ability_primacy = XENO_PRIMARY_ACTION_5
+	plasma_cost = 20
+	xeno_cooldown = 17 SECONDS
+
+	var/damage_threshold = 140 // Requires 140 damage taken within 6 seconds to activate the ability
+	var/damage_accumulated = 0 // Intially zero, gets damage added when the ability is activated
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_abilities.dm:63]:
```suggestion
	/// Requires 140 damage taken within 6 seconds to activate the ability
	var/damage_threshold = 140
	/// Initially zero, gets damage added when the ability is activated
	var/damage_accumulated = 0
```

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 2]
```diff
@@ -262,3 +262,73 @@
...
+
+	steelcrest.balloon_alert(steelcrest, "begins to take in oncoming damage!")
+
+	to_chat(steelcrest, SPAN_XENONOTICE("You begin to take in oncoming damage!"))
+
+	steelcrest.add_filter("steelcrest_enraging", 1, list("type" = "outline", "color" = "#421313", "size" = 1))
+
+	apply_cooldown()
+	..()
+	return
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:289]:
```suggestion
	return ..()
```

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 3]
```diff
@@ -262,3 +262,73 @@
...
+
+
+/datum/action/xeno_action/onclick/soak/proc/damage_accumulate(owner, damage_data, damage_type)
+	SIGNAL_HANDLER
+
+	var/damage_taken = damage_data["damage"]
+
+	damage_accumulated += damage_taken
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:297]:
```suggestion
	damage_accumulated += damage_data["damage"]
```

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 4]
```diff
@@ -262,3 +262,73 @@
...
+
+	if(damage_accumulated >= damage_threshold)
+		addtimer(CALLBACK(src, PROC_REF(enraged), owner))
+		UnregisterSignal(owner, COMSIG_XENO_TAKE_DAMAGE) // Two Unregistersignal because if the enrage proc doesnt happen, then it needs to stop counting
+
+/datum/action/xeno_action/onclick/soak/proc/stop_accumulating()
+	UnregisterSignal(owner, COMSIG_XENO_TAKE_DAMAGE)
+
+	damage_accumulated = 0
+	to_chat(src, SPAN_XENONOTICE("You stop taking in oncoming damage."))
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:307]: send this to the owner, src is the datum

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 5]
```diff
@@ -262,3 +262,73 @@
...
+	enraged_mob.gain_health(75) // pretty reasonable amount of health recovered
+
+	// Check actions list for tail slam and reset it's cooldown if it's there
+	var/datum/action/xeno_action/activable/tail_stab/slam/slam_action = null
+	for (var/datum/action/xeno_action/activable/tail_stab/slam/slam in owner.actions)
+		slam_action = slam
+		break
+
+	if (slam_action && !slam_action.action_cooldown_check())
+		slam_action.end_cooldown()
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:326]:
```suggestion
	var/datum/action/xeno_action/activable/tail_stab/slam/slam_action = locate() in owner.actions
	
	if (slam_action && !slam_action.action_cooldown_check())
		slam_action.end_cooldown()
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — APPROVED** [review]

**Review by Agameofscones — CHANGES_REQUESTED** [review]
the ability icon here is nearly identical to the pre-existing crest_defense.

It needs to be differentiated more heavily.

**Review by morrowwolf — COMMENTED** [review]
Looks cool balance-wise, let's see how it goes in testing though. Get the sprite changes sorted. In the meantime I'll testmerge and keep an eye on it.

**Review by morrowwolf — APPROVED** [review]
Balance seems fine and was functional. Maybe one day we'll get people to play steelcrest.

Sort the sprite stuff and we're good to go.

**Review by sleepynecrons — CHANGES_REQUESTED** [review]
The new icon is better but the glowing effect is a bit messy. Try something like this 

5yjdHDT

You could also make the glowing effect a quarter-circle instead of following the silhouette of the head. It just needs to be a bit cleaner.

image
(You can copypaste this one into whatever spriting software you're using if you want to use it)

**Review by sleepynecrons — APPROVED** [review]
icon looks good now 👍

**Review by harryob — APPROVED** [review]
code looks fine

**Pull request merged**